internal/federation/handle.go

```diff
@@ -277,3 +279,101 @@ func HandleMediaRequests(mediaIds map[string]func(w http.ResponseWriter)) func(*
+func HandleTransactionRequests() func(*Server) {
+	return func(srv *Server) {
+		srv.mux.Handle("/_matrix/federation/v1/send/{transactionID}", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			// Check federation signature
+			fedReq, errResp := gomatrixserverlib.VerifyHTTPRequest(
+				req, time.Now(), gomatrixserverlib.ServerName(srv.ServerName), srv.keyRing,
+			)
+			if fedReq == nil {
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Read the request body
+			body, err := ioutil.ReadAll(req.Body)
+			if err != nil {
+				errResp := util.MessageResponse(400, err.Error())
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Unmarshal the request body into a transaction object
+			var transaction gomatrixserverlib.Transaction
+			err = json.Unmarshal(body, &transaction)
+			if err != nil {
+				errResp := util.MessageResponse(400, err.Error())
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Extract the transaction ID from the request vars
+			vars := mux.Vars(req)
+			transaction.TransactionID = gomatrixserverlib.TransactionID(vars["transactionID"])
+
+			// Transactions are limited in size; they can have at most 50 PDUs and 100 EDUs.
+			// https://matrix.org/docs/spec/server_server/latest#transactions
+			if len(transaction.PDUs) > 50 || len(transaction.EDUs) > 100 {
+				errResp := util.MessageResponse(400, "Transactions are limited to 50 PDUs and 100 EDUs")
+				w.WriteHeader(errResp.Code)
+				b, _ := json.Marshal(errResp.JSON)
+				w.Write(b)
+				return
+			}
+
+			// Construct a response and fill as we process each PDU
+			response := gomatrixserverlib.RespSend{}
+			response.PDUs = make(map[string]gomatrixserverlib.PDUResult, 0)
+			for _, pdu := range transaction.PDUs {
+				var header struct {
+					RoomID string `json:"room_id"`
+				}
+				if err := json.Unmarshal(pdu, &header); err != nil {
+					log.Printf("Transaction '%s': Failed to extract room ID from event", transaction.TransactionID)
+					// We don't know the event ID at this point so we can't return the
+					// failure in the PDU results
+				}
+
+				// Retrieve the room version from the server
+				room := srv.rooms[header.RoomID]
+				roomVersion := gomatrixserverlib.RoomVersion(room.Version)
+
+				event, err := gomatrixserverlib.NewEventFromUntrustedJSON(pdu, roomVersion)
+				if err != nil {
+					// We were unable to verify or process this event.
+					// Add this PDU to the response along with an error message.
+					response.PDUs[event.EventID()] = gomatrixserverlib.PDUResult{
+						Error: err.Error(),
+					}
+					continue
+				}
+
+				// Store this PDU in the room's timeline
+				room.Timeline = append(room.Timeline, &event)
```

Comment:
Need to update state too. Just call https://github.com/matrix-org/complement/blob/master/internal/federation/server_room.go#L23